### PR TITLE
Support for uploading qlog artifacts

### PR DESCRIPTION
**Pull request by hlandau**

### CODE REVIEW

**Review by levitte — CHANGES_REQUESTED** [review]
Looking at the test recipe, I'm not sure why the `test_quic_multistream` has to be so dependent on if we run a CI or not.  Like, the verification against `verify-qlog.py`, is there a reason it shouldn't run when I'm test locally on my laptop?

File: test/recipes/70-test_quic_multistream.t
Comment (levitte) [inline test/recipes/70-test_quic_multistream.t:61]: This strikes me as very odd.  I would rather have the caller (i.e. the CI) be able to set the environment variable `QLOGDIR`, and let the caller do the rsync.

You could have this at the start of the recipe:

```
$ENV{QLOGDIR} = result_dir("qlog-output") unless $ENV{QLOGDIR};
```

and then go from there.

**Comment by hlandau** [comment]
It would depend on Python.

File: test/recipes/70-test_quic_multistream.t
Comment (hlandau) [reply test/recipes/70-test_quic_multistream.t:61]: Hmm... but how do we get `result_dir()`? That is computed within the test framework, and it's where unit test-specific output is supposed to go.

**Comment by levitte** [comment]
> It would depend on Python.

Then perhaps it should actually depend on if python is available or not.  The module `IPC::Cmd` has a functions `can_run` for that sort of thing

**Comment by hlandau** [comment]
> > It would depend on Python.
> 
> Then perhaps it should actually depend on if python is available or not. The module `IPC::Cmd` has a functions `can_run` for that sort of thing

We use that for probing in Configure which is expected. It seems... rather "surprising" for us to vary how much of our unit test suite based on random environmental factors in this way...

**Comment by levitte** [comment]
> > > It would depend on Python.
> > 
> > 
> > Then perhaps it should actually depend on if python is available or not. The module `IPC::Cmd` has a functions `can_run` for that sort of thing
> 
> We use that for probing in Configure which is expected. It seems... rather "surprising" for us to vary how much of our unit test suite based on random environmental factors in this way...

...

https://github.com/openssl/openssl/blob/925118e8c3b1041ce7f9840c2d67e7f878123e6b/test/recipes/01-test_symbol_presence.t#L27-L28

It's not like we don't have previous experience 😉

**Comment by hlandau** [comment]
This is now working. Artifacts: https://github.com/openssl/openssl/actions/runs/7871749345?pr=23551

**Comment by hlandau** [comment]
I've moved some of the artifacts upload code out of the unit test recipe.

The QLOGDIR code should definitely remain in the recipe; test-runs is exactly where this data is supposed to go and it is useful data even if not running in CI.

The qlog Python CI test was already merged so I don't think enabling it by default should be discussed in this PR. IMO a new issue should be opened regarding this.

**Review by t8m — APPROVED** [review]

**Comment by nhorman** [comment]
I think this can be closed, as https://github.com/openssl/openssl/pull/23715 has been merged.  We now will get the qlog json output dumped to the ci log in the event that qlog testing fails

**Comment by hlandau** [comment]
Err no. This is separate and much more useful

**Comment by hlandau** [comment]
Still needs a second review! @nhorman

**Review by t-j-h — APPROVED** [review]

**Comment by hlandau** [comment]
@nhorman Do you want this for 3.3?

It is pure CI so IMO it is acceptable post-alpha. We are going to be maintaining this branch for some time and having coverage here would be desirable. So I would support this. Your call.

**Review by nhorman — APPROVED** [review]

**Comment by nhorman** [comment]
I'd be ok with the addition of this, but for clarification, I think this would generally be an RSC decision, no?

**Comment by nhorman** [comment]
also @hlandau given https://github.com/openssl/openssl/pull/23715 is merged, is this a high priority?  That is to say, while I'm not opposed to this at all, a failure in verifying qlog output dumps the json directlly to the console now as well for review.

**Comment by nhorman** [comment]
This item is likely small enough that I'll just make the call, this can go in (#23715 notwithstanding)

**Comment by hlandau** [comment]
Merged to master and cherry picked to 3.3.

**Pull request closed without merging**
